vending-app/src/main/kotlin/com/android/vending/installer/Install.kt

```diff
@@ -30,127 +33,193 @@ import java.io.OutputStream
-
+        Log.e(TAG, "installPackagesInternal pendingSessions size: ${SessionResultReceiver.pendingSessions.size}")
         SessionResultReceiver.pendingSessions[sessionId] = SessionResultReceiver.OnResult(
-            onSuccess = {
-                deferred.complete(Unit)
-                emitProgress(sessionId, InstallComplete)
-                        },
-            onFailure = { message ->
-                deferred.completeExceptionally(RuntimeException(message))
-                emitProgress(sessionId, InstallError(message ?: "UNKNOWN"))
-            }
+                onSuccess = {
+                    deferred.complete(Unit)
+                    emitProgress(notifyId, InstallComplete)
+                },
+                onFailure = { message ->
+                    deferred.completeExceptionally(RuntimeException(message))
+                    emitProgress(notifyId, InstallError(message ?: "UNKNOWN"))
+                }
         )
-
         val intent = Intent(this, SessionResultReceiver::class.java)
-        val pendingIntent = PendingIntent.getBroadcast(this, sessionId, intent,
-            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE)
+        intent.putExtra(SessionResultReceiver.KEY_NOTIFY_ID, notifyId)
+        intent.putExtra(SessionResultReceiver.KEY_PACKAGE_NAME, packageName)
+        val pendingIntent = PendingIntent.getBroadcast(
+                this, sessionId, intent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
+        )
 
-        emitProgress(sessionId, CommitingSession)
+        emitProgress(notifyId, CommitingSession)
         session.commit(pendingIntent.intentSender)
         // don't abandon if `finally` step is reached after this point
-        session = null
+        //session.close()
 
         Log.d(TAG, "installPackages session commit")
         return deferred.await()
-    } catch (e: IOException) {
-        Log.e(TAG, "Error installing packages", e)
-        emitProgress(sessionId, InstallError(e.message ?: "UNKNOWN"))
+    } catch (e: Exception) {
+        Log.w(TAG, "Error installing packages", e)
+        emitProgress(notifyId, InstallError(e.message ?: "UNKNOWN"))
         throw e
     } finally {
-        // discard downloaded data
+        // Close the session to release resources after error
         session?.let {
-            Log.d(TAG, "Discarding session after error")
-            it.abandon()
+            Log.d(TAG, "Error occurred, session cleanup may be required")
+            it.close()
         }
     }
 }
+
+private fun Context.computeUniqueKey(packageName: String, componentNames: List<String>) : String {
+    val versionCode = PackageInfoCompat.getLongVersionCode (
+            packageManager.getPackageInfo(packageName, 0)
+    )
+    val key = componentNames.joinToString(separator = "_", prefix = "${packageName}_${versionCode}")
+    Log.d(TAG, "computeUniqueKey: $key")
+    return key
+}
```

Comment:
This function doesn't work for packages that are not yet installed, instead throwing `PackageManager$NameNotFoundException`. Please handle this case – from my tests it works to simply not consider the version code for the key in that case.